feat(root): include local tunnel work

denis-kralj-novu opened the pull request:
### What changed? Why was the change needed?

The `create-novu-app` package now has the ability to open a localtunnel to your app on `npm run dev`

For local testing of behavior, the project can be built with `npm run dev`. It can then be triggered via `npm run start -- --api-key=[YOUR_API_KEY]`. using pnpm is slightly different with `pnpm run start --api-key=[YOUR_API_KEY]`

### With enterprise
- on user registration, the bridge endpoint url is also stored in the environment configuration
- the bridge endpoint url gets updated when the development environment api key is regenerated
- the environments endpoint allows to update the bridge endpoint url on PUT
## Related enterprise PR
This PR depends on [this enterprise PR](https://github.com/novuhq/packages-enterprise/pull/139)

denis-kralj-novu committed:
feat: include local tunnel work

Add local tunnel script to novu app
Add storage of local tunnel url

denis-kralj-novu commented on apps/api/src/app/environments/usecases/update-environment/update-environment.usecase.ts:
```diff
@@ -32,4 +41,24 @@ export class UpdateEnvironment {
     );
   }
+  async shouldUpdateEchoConfiguration(command: UpdateEnvironmentCommand): Promise<boolean> {
+    if (process.env.NOVU_ENTERPRISE === 'true' || process.env.CI_EE_TEST === 'true') {
```
I initially thought that I could use the command that I defined in the ee package for updating the endpoint. However, this would then lead to a situation where that command would return the results of the update, wheres the previous set of adjustments in the caller would have its own result. I could do consolidation logic since it is aiming via `environmentId`, but that seemed a bit messy. Let me know if that would still be preferable to this way of updating the data.

denis-kralj-novu replied:
@rifont let me hear your thoughts on this as well

rifont replied:
I think this is a suitable approach for the time being 👍

denis-kralj-novu commented on apps/web/package.json:
```diff
@@ -5,7 +5,7 @@
     "start": "pnpm panda --watch & cross-env NODE_OPTIONS=--max_old_space_size=8192 PORT=4200 react-app-rewired start",
     "prebuild": "rimraf build",
-    "build": "pnpm panda && cross-env NODE_OPTIONS=--max_old_space_size=4096 GENERATE_SOURCEMAP=false react-app-rewired --max_old_space_size=4096 build",
+    "build": "pnpm panda && cross-env NODE_OPTIONS=--max_old_space_size=8192 GENERATE_SOURCEMAP=false react-app-rewired --max_old_space_size=8192 build",
```
Without this the build would fail with the following error:
```
FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
```
Since this setting is at the value of `8192` in the start command, I adjusted it here and then the build passed.

denis-kralj-novu commented on package.json:
```diff
@@ -13,7 +13,7 @@
     "commit": "cz",
     "nx": "nx",
-    "lint-staged": "lint-staged",
+    "lint-staged": "NODE_OPTIONS=--max_old_space_size=8192 lint-staged",
```
Without this committing would fail with the following error:
```
FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
```
I aligned this value with the ones present in `apps/web`

rifont replied:
This could be an indication of a memory leak, we shouldn't need this much memory to run a linting job. In the past when this has occurred, restarting my machine fixes the issue. I'm hesitant to make this change because it may mask further memory leaks. I'd suggest we revert for now.

```suggestion
    "lint-staged": "lint-staged",
```

denis-kralj-novu replied:
reverted back

denis-kralj-novu commented on packages/create-novu-app/index.ts:
```diff
@@ -19,6 +19,8 @@ let projectPath = '';
 
 const handleSigTerm = () => process.exit(0);
 
+const defaultTunnelHost = 'https://localtunnel.me';
```
for some reason the default value in the option doesn't work so I needed to place it here to have access to it.

denis-kralj-novu committed:
feat: include local tunnel work

Add test for user registration case in enterprise

denis-kralj-novu committed:
feat: include local tunnel work

Consolidate scripts to one source

denis-kralj-novu committed:
feat: include local tunnel work

Update modules to be typescript instead of javascript
Update output package json command
Update build command to clean dist before building

denis-kralj-novu committed:
feat: include local tunnel work

Update formatting
Add tunnel interface to conform to

rifont commented on packages/create-novu-app/templates/index.ts:
```diff
@@ -166,6 +173,21 @@ export const installTemplate = async ({
+  await fs.writeFile(
+    path.join(root, ".env"),
+    `PORT=${port}` +
+      os.EOL +
+      `TUNNEL_SUBDOMAIN=${subdomain}` +
+      os.EOL +
+      `TUNNEL_HOST=${host}` +
+      os.EOL,
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
💭 **thought (non-blocking):** Consider `Object.entries(obj).reduce(...)` with `obj` being a map of env var key to value.‏

denis-kralj-novu replied:
done

rifont commented on packages/create-novu-app/templates/index.ts:
```diff
@@ -175,7 +197,10 @@ export const installTemplate = async ({
+      tunnel:
+        "node --loader ts-node/esm --no-warnings=ExperimentalWarning scripts/tunnel.mts",
+      "next-dev": `next dev --port=${port}`,
+      dev: "concurrently 'npm run tunnel' 'npm run next-dev'",
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
⚠ **issue:** It feels a bit heavy-handed and opinionated to add a library to concurrently run the tunnel with Dev server. Are there `next dev` scripts we can hook into to setup the tunnel in a more transparent way?‏

denis-kralj-novu replied:
discussed 1on1, decided to go with `concurrently`

rifont commented on packages/create-novu-app/index.ts:
```diff
@@ -41,6 +43,22 @@ const program = new Commander.Command(packageJson.name)
+`
+  )
+  .option(
+    '-th, --tunnel-host',
```
I find it a bit non-standard to have a shorthand flag variant with 2 characters. What do we think of shortening to `-t`? This interface shouldn't change very frequently, if we have another `t` character we can find something else. It's better to optimize for what we have today given that this is a critical part of the onboarding flow right now.

```suggestion
    '-t, --tunnel-host',
```

denis-kralj-novu replied:
Yep, that makes sense :)

rifont commented on libs/application-generic/src/utils/buildBridgeEndpointUrl.ts:
```diff
@@ -0,0 +1,19 @@
+ */
+export const buildBridgeSubdomain = (apiKey: string): string => {
+  return createHash('md5').update(apiKey).digest('hex');
+};
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
⚠ **issue:** This hashing function is simple enough for us to duplicate across both places we need it, adding a dependency on `@novu/application-generic` will significantly bloat the bundle size of `create-novu-app`. Let's duplicate for now until we have a better home for these types of functions, accepting double maintenance cost.‏

denis-kralj-novu replied:
Right, agree, should I add a note somewhere to say that there should be parity between the 2 places?

rifont commented on packages/create-novu-app/templates/app-react-email/ts/scripts/tunnel.mjs:
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
⚠ **issue:** My thinking is that we would generate the tunnel URL at runtime on the `create-novu-app` client to support the use-case of rolling the API key to a new value. Currently, if the developer rolls the API-key ("rolls" meaning they invalidate the old key and generate a new one), the tunnel URL becomes invalid. We want to transparently support providing a new API key without needing to regenerate the tunnel.‏

denis-kralj-novu replied:
Right, the subdomain gets calculated when `create-novu-app` is run, so we have that today. The scenario with the rolling, I'm thinking to just add the regeneration of the url as well to the process of rolling the API key as well. The roll use-case would call the regenerate URL use-case, what do you think?

denis-kralj-novu replied:
subdomain now generated based on api key on `npm run dev`

rifont commented on packages/create-novu-app/templates/app/ts/scripts/tunnel.mjs:
```diff
@@ -0,0 +1,34 @@
+    Number.isNaN(parseInt(process.env.PORT))
+  ) {
+    throw new Error(
+      "POST environment variable value is required and should be a positive integer",
```
Typo
```suggestion
      "PORT environment variable value is required and should be a positive integer",
```

denis-kralj-novu committed:
feat: include local tunnel work

Adjust per PR comments and discussion

denis-kralj-novu committed:
feat: include local tunnel work
Subdomain calculated from ApiKey on launch instead on scaffold
Run script with tsx

denis-kralj-novu committed:
feat: include local tunnel work

Adjust how concurrently handles tasks
Adjust text in output

denis-kralj-novu committed:
feat: include local tunnel work

Regenerate echo url on api key rotation

denis-kralj-novu committed:
feat: include local tunnel work

Remove unused package

denis-kralj-novu committed:
feat: include local tunnel work

Update lock file

denis-kralj-novu committed:
Merge branch 'next' into include-local-tunnel

denis-kralj-novu committed:
feat: include local tunnel work

Update lock file

denis-kralj-novu committed:
feat: include local tunnel work

Update lock file

denis-kralj-novu committed:
feat: include local tunnel work

Update lock file

denis-kralj-novu committed:
Merge branch 'next' into include-local-tunnel

denis-kralj-novu commented on apps/api/src/app/environments/usecases/update-environment/update-environment.usecase.ts:
```diff
@@ -20,10 +20,19 @@ export class UpdateEnvironment {
     }
 
-    if (command.dns && command.dns.inboundParseDomain !== '') {
+    if (command.dns && command.dns.inboundParseDomain && command.dns.inboundParseDomain !== '') {
```
Added check here to also ensure that the value isn't falsy

rifont replied:
🚀

denis-kralj-novu committed:
feat: include local tunnel work

PR suggestions

denis-kralj-novu committed:
Merge branch 'next' into include-local-tunnel

denis-kralj-novu committed:
feat: include local tunnel work

Update restart cooldown to be 1sec

denis-kralj-novu committed:
Merge branch 'next' into include-local-tunnel

denis-kralj-novu committed:
feat: include local tunnel work

Add error handler logic that enables restart on tunnel fail

rifont commented on libs/application-generic/src/utils/buildBridgeEndpointUrl.ts:
```diff
@@ -0,0 +1,22 @@
+};
+
+/*
+ * Creates a bridge subdomain based on the apiKey provided. This function is used in several
+ * places, including packages/create-novu-app/templates/index.ts when generating the
+ * subdomain in the bridge application. Developers should take care to keep changes
+ * in sync.
+ */
```
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise.  -->
👏 **praise:** ‏Nice comment 🙏

rifont commented on packages/create-novu-app/templates/index.ts:
```diff
@@ -175,7 +193,9 @@ export const installTemplate = async ({
-      dev: "next dev --port=4000",
+      tunnel: "tsx scripts/tunnel.mts",
+      "next-dev": `next dev --port=${port}`,
+      dev: 'concurrently -k --restart-tries 5 --restart-after 1000 --names "📡 TUNNEL,🖥️  SERVER" -c "bgBlue.bold,bgMagenta.bold" "npm:tunnel" "npm:next-dev"',
```
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise.  -->
👏 **praise:** 😍 love the emojis‏

rifont reviewed (approved):
Awesome work! It's a very clean DX. Let's ship it! 🚀

denis-kralj-novu committed:
feat: include local tunnel work

update ee package reference
